test/distributed/fsdp/test_fsdp_hybrid_shard.py

```diff
@@ -0,0 +1,245 @@
+                self.assertEqual(model.lin1.process_group, intra_pg)
+                self.assertEqual(model.lin1._inter_node_pg, inter_pg)
+                model = FSDP(model, process_group=intra_pg)
+                inp = torch.randn(4, 10)
+                # Errors during _lazy_init
+                with self.assertRaisesRegex(ValueError, "expect sharding strategies to be the same"):
+                    model(inp)
+
+    @skip_if_lt_x_gpu(2)
+    def test_hybrid_shard_pg_mismatch_raises(self):
+        model = MyModel().cuda()
+        intra_pg = self.process_group
+        inter_pg = dist.new_group(ranks=[self.rank])
+        # Mismatched process groups for intra-node
+        model.lin1 = FSDP(
+            model.lin1, process_group=(intra_pg, inter_pg), sharding_strategy=ShardingStrategy.HYBRID_SHARD
+        )
+        model = FSDP(
+            model, process_group=(dist.new_group(), dist.new_group()), sharding_strategy=ShardingStrategy.HYBRID_SHARD
+        )
+        # Errors during lazy_init
+        inp = torch.randn(4, 10)
+        with self.assertRaisesRegex(ValueError, "intra-node process groups do not match"):
+            model(inp)
+
+        # Mismatched process groups for inter-node
+        model = MyModel().cuda()
+        model.lin1 = FSDP(
+            model.lin1, process_group=(intra_pg, inter_pg), sharding_strategy=ShardingStrategy.HYBRID_SHARD
+        )
+        model = FSDP(
+            model, process_group=(intra_pg, dist.new_group()), sharding_strategy=ShardingStrategy.HYBRID_SHARD
+        )
+        with self.assertRaisesRegex(ValueError, "inter-node process groups do not match"):
+            model(inp)
+
+    @skip_if_lt_x_gpu(2)
+    def test_invalid_pg_specification_raises(self):
+        pol = ModuleWrapPolicy({nn.Linear})
+        model = MyModel().cuda()
+        with self.assertRaisesRegex(
+            ValueError,
+            "Expected process_group to be passed in"
+        ):
+            model = FSDP(
+                model,
+                auto_wrap_policy=pol,
+                process_group=self.process_group,
+                sharding_strategy=ShardingStrategy.HYBRID_SHARD
+            )
+
+    # TODO - add test for Zero-2 style sharding ensure params are not
+    # resharded after forward.
+
+    @skip_if_lt_x_gpu(2)
+    def test_fsdp_hybrid_shard_basic_setup(self):
+        """
+        Tests basic functionality of HYBRID_SHARD and _HYBRID_SHARD_ZERO2:
+            1. Inter and intra-node process groups are correctly setup
+            2. Process groups are the same across FSDP wrapped units
+            3. reduce_scatter and allreduce called the expected no. of times
+        """
+        for sharding_strategy in [
+            ShardingStrategy.HYBRID_SHARD, ShardingStrategy._HYBRID_SHARD_ZERO2
+        ]:
+            with self.subTest(sharding_strategy=sharding_strategy):
```

Comment:
Eventually, we should refactor use `self.run_subtests()` which wraps over this `self.subTest` and includes the final `dist.barrier()`.